src/integrationtests/java/com/aws/greengrass/integrationtests/logmanager/LogManagerTest.java

```diff
@@ -453,4 +466,162 @@ void GIVEN_system_config_with_small_periodic_interval_WHEN_active_logs_included_
+        logFiles.sort(Comparator.comparingLong(LogFile::lastModified));
+        for (int i = 0; i < logFiles.size(); i++) {
+            LogFile logFile = logFiles.get(i);
+            String randomName2 = UUID.randomUUID().toString();
+            LogFile renameFile = new LogFile(Files.createTempFile(tempDirectoryPath,
+                    randomName2 + "UserComponentB.", "").toUri());
+            String tempFileHash = logFile.hashString();
+            logFile.renameTo(renameFile);
+            // add more data into the active file.
+            if (i == logFiles.size() - 1) {
+                //generate 10250 bytes
+                List<String> randomMessages = generateRandomMessages();
+                for (String messageBytes : randomMessages) {
+                    addDataToFile(messageBytes, renameFile.toPath());
+                }
+            }
+
+            assertEquals(renameFile.hashString(), tempFileHash);
+            assertFalse(logFile.exists());
+        }
+
+        assertTrue(isActiveFileProcessedOrTimeout(logManagerService));
+
+        // All files are renamed, and only the active file adds 10250 bytes contents, so we shall expect two tries
+        // for putLogEvents, and one is for uploading 4 files content, another is for only upload 10250 bytes.
+        verify(cloudWatchLogsClient, times(2)).putLogEvents(captor.capture());
+
+        List<PutLogEventsRequest> putLogEventsRequests = captor.getAllValues();
+        assertEquals(2, putLogEventsRequests.size());
+        for (int i=0; i < 2; i++) {
+            PutLogEventsRequest request = putLogEventsRequests.get(i);
+            assertEquals(calculateLogStreamName(THING_NAME, LOCAL_DEPLOYMENT_GROUP_NAME), request.logStreamName());
+            assertEquals("/aws/greengrass/UserComponent/" + AWS_REGION + "/UserComponentB", request.logGroupName());
+            assertNotNull(request.logEvents());
+            if (i == 1) {
+                assertEquals(DEFAULT_LOG_LINE_IN_FILE, request.logEvents().size());
+                assertEquals(DEFAULT_FILE_SIZE,
+                        request.logEvents().stream().mapToLong(value -> value.message().length()).sum());
+            }
+            if (i == 0) {
+                assertEquals(DEFAULT_LOG_LINE_IN_FILE * testFileNumber, request.logEvents().size());
+                assertEquals(DEFAULT_FILE_SIZE * testFileNumber,
+                        request.logEvents().stream().mapToLong(value -> value.message().length()).sum());
+            }
+        }
+        // confirm no file is deleted
+        Pattern logFileNamePattern = Pattern.compile("\\w*UserComponentB\\w*.\\w*");
+        LogFileGroup logFileGroup = LogFileGroup.create(logFileNamePattern, tempDirectoryPath.toUri(), mockInstant);
+        assertEquals(testFileNumber, logFileGroup.getLogFiles().size());
+    }
+
+    boolean isActiveFileProcessedOrTimeout(LogManagerService logManagerService) throws InterruptedException {
```

Comment:
The idea of having that observer pattern that emit event was so that we avoid things like this, you could do something like


```suggestion
    private Runnable subscribeToActiveFileProcessed(LogManagerService service) throws InterruptedException {
       CountDownLatch latch = new CountDownLatch(1);
        Runnable deregister = service.registerListener((Event event) -> {
              if (event == Event.ACTIVE_FILE_PROCESSED) {
                     latch.countDown();
              }
        })
        
        return () -> {
             latch.await(15, TimeUnit.SECONDS);
             deregister();
        }
   }
   
   // Then on the test
   @Test
   void fooBar() {
       // this just registers a latch and a listener
       Runnable waitForActiveFileToBeProcessed = subscribeToActiveFileProcessed(service);
       
       // Do some stuff
       waitForActiveFileToBeProcessed(); // <-- calling this will block the thread until the latch is released
   }
```